pkg/migration/controllers/migration.go

```diff
@@ -0,0 +1,667 @@
+	object runtime.Unstructured,
+) (runtime.Unstructured, error) {
+	spec, err := collections.GetMap(object.UnstructuredContent(), "spec")
+	if err != nil {
+		return nil, err
+	}
+	delete(spec, "claimRef")
+	delete(spec, "storageClassName")
+
+	return m.Driver.UpdateMigratedPersistentVolumeSpec(object)
+}
+
+func (m *MigrationController) prepareApplicationResource(
+	migration *storkv1.Migration,
+	object runtime.Unstructured,
+) (runtime.Unstructured, error) {
+	if migration.Spec.StartApplications {
+		return object, nil
+	}
+
+	// Reset the replicas to 0 and store the current replicas in an annotation
+	content := object.UnstructuredContent()
+	spec, err := collections.GetMap(content, "spec")
+	if err != nil {
+		return nil, err
+	}
+	replicas := spec["replicas"].(int64)
+	annotations, err := collections.GetMap(content, "metadata.annotations")
+	annotations[storkMigrationReplicasAnnotation] = strconv.FormatInt(replicas, 10)
+	spec["replicas"] = 0
+	return object, nil
+}
+
+func (m *MigrationController) applyResources(
+	migration *storkv1.Migration,
+	objects []runtime.Unstructured,
+) error {
+	remoteConfig, err := getClusterPairSchedulerConfig(migration.Spec.ClusterPair)
+	if err != nil {
+		return err
+	}
+
+	client, err := kubernetes.NewForConfig(remoteConfig)
+	if err != nil {
+		return err
+	}
+
+	// First make sure all the namespaces are created on the
+	// remote cluster
+	for _, ns := range migration.Spec.Namespaces {
+		namespace, err := k8s.Instance().GetNamespace(ns)
+		if err != nil {
+			return err
+		}
+		_, err = client.CoreV1().Namespaces().Create(&v1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:   namespace.Name,
+				Labels: namespace.Labels,
+			},
+		})
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+	}
+
+	remoteDynamicClientPool := dynamic.NewDynamicClientPool(remoteConfig)
+	for _, o := range objects {
+		dynamicClient, err := remoteDynamicClientPool.ClientForGroupVersionKind(o.GetObjectKind().GroupVersionKind())
+		if err != nil {
+			return err
+		}
+		metadata, err := meta.Accessor(o)
+		if err != nil {
+			return err
+		}
+		objectType, err := meta.TypeAccessor(o)
+		if err != nil {
+			return err
+		}
+		logrus.Infof("Applying %v %v", objectType.GetKind(), metadata.GetName())
+		resource := &metav1.APIResource{
+			Name:       strings.ToLower(objectType.GetKind()) + "s",
+			Namespaced: len(metadata.GetNamespace()) > 0,
+		}
+		client := dynamicClient.Resource(resource, metadata.GetNamespace())
+		unstructured, ok := o.(*unstructured.Unstructured)
+		if !ok {
+			return fmt.Errorf("Unable to cast object to unstructured: %v", o)
+		}
+		_, err = client.Create(unstructured)
+		if err != nil && apierrors.IsAlreadyExists(err) {
+			switch objectType.GetKind() {
+			// Don't want to delete the Volume resources
+			case "PersistentVolumeClaim", "PersistentVolume":
+				err = nil
+			default:
+				// Delete the resource if it already exists on the destination
```

Comment:
we should let the users override this behavior from the migration spec. The default can be current behavior to override.

This might end up overriding a config map or secret which might be intended.

Reply:
Will add as a separate change